opm/simulators/linalg/AmgxPreconditioner.hpp

```diff
@@ -0,0 +1,348 @@
+        AMGX_SAFE_CALL(AMGX_matrix_upload_all(A_amgx_, N_, nnz_, block_size, block_size,
+                                             row_ptrs.data(), col_indices.data(),
+                                             values, nullptr));
+        update();
+    }
+
+    /**
+     * @brief Destructor for the AmgxPreconditioner class.
+     *
+     * Cleans up resources allocated by the preconditioner.
+     */
+    ~AmgxPreconditioner()
+    {
+        const matrix_field_type* values = &(A_[0][0][0][0]);
+        AMGX_SAFE_CALL(AMGX_unpin_memory(const_cast<matrix_field_type*>(values)));
+        if (solver_) {
+            AMGX_SAFE_CALL(AMGX_solver_destroy(solver_));
+        }
+        if (x_amgx_) {
+            AMGX_SAFE_CALL(AMGX_vector_destroy(x_amgx_));
+        }
+        if (b_amgx_) {
+            AMGX_SAFE_CALL(AMGX_vector_destroy(b_amgx_));
+        }
+        if (A_amgx_) {
+            AMGX_SAFE_CALL(AMGX_matrix_destroy(A_amgx_));
+        }
+        // Destroying resrouces and config crashes when reinitializing
+        //if (rsrc_) {
+        //    AMGX_SAFE_CALL(AMGX_resources_destroy(rsrc_));
+        //}
+        //if (cfg_) {
+        //    AMGX_SAFE_CALL(AMGX_config_destroy(cfg_));
+        //}
+    }
+
+    /**
+     * @brief Pre-processing step before applying the preconditioner.
+     *
+     * This method is currently a no-op.
+     *
+     * @param v The update vector.
+     * @param d The defect vector.
+     */
+    void pre(X& /*v*/, Y& /*d*/) override {
+    }
+
+    /**
+     * @brief Applies the preconditioner to a vector.
+     *
+     * Performs one AMG cycle to solve the system.
+     * Involves uploading vectors to AMGX, applying the preconditioner, and downloading the result.
+     *
+     * @param v The update vector.
+     * @param d The defect vector.
+     */
+    void apply(X& v, const Y& d) override
+    {
+        OPM_TIMEBLOCK(prec_apply);
+
+        // Upload vectors to AMGX
+        AMGX_SAFE_CALL(AMGX_vector_upload(x_amgx_, N_, block_size, &v[0][0]));
+        AMGX_SAFE_CALL(AMGX_vector_upload(b_amgx_, N_, block_size, &d[0][0]));
+
+        // Apply preconditioner
+        AMGX_SAFE_CALL(AMGX_solver_solve(solver_, b_amgx_, x_amgx_));
+
+        // Download result
+        AMGX_SAFE_CALL(AMGX_vector_download(x_amgx_, &v[0][0]));
+    }
+
+    /**
+     * @brief Post-processing step after applying the preconditioner.
+     *
+     * This method is currently a no-op.
+     *
+     * @param v The update vector.
+     */
+    void post(X& /*v*/) override {
+    }
+
+    /**
+     * @brief Updates the preconditioner with the current matrix values.
+     *
+     * This method should be called whenever the matrix values change.
+     */
+    void update() override
+    {
+        OPM_TIMEBLOCK(prec_update);
+        copyMatrixToAmgx();
+        if (update_counter_ == 0) {
+            AMGX_SAFE_CALL(AMGX_solver_setup(solver_, A_amgx_));
```

Comment:
Is it safe to do setup without destroying it first?

Reply:
Yes this should be safe. There is some functionality to reuse some memory and levels of the AMG in the setup function also. I have done some digging in the code but still not 100 % clear to me how much more is reused with resetup.

Only using the setup function is actually the way to do it with the official API and in the manual, as the resetup was marked deprecated in the initial AMGX commit 7 years ago... https://github.com/NVIDIA/AMGX/blob/69e625e99b8788ac7b259c6b217ad4453cc69464/examples/amgx_capi.h#L522
But since it still available 7 years later and important to us I decided to still use it.